### PR TITLE
Authorization with cloud & publisher portal

### PR DESCRIPTION
**Pull request by graywolf336**
We are enabling publishers to authenticate with our Rocket.Chat Cloud. We have also been working on a publisher portal so that publishers can easily interact with their apps outside of just the CLI. In a few days more information will be released about it and we will be contacting existing publishers and then enabling new publishers to sign up.

### CODE REVIEW

File: src/misc/cloudAuth.ts [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+
+            throw err;
+        }
+    }
+
+    private buildAuthorizeUrl(codeChallenge: string) {
+        const data = {
+            client_id: '5d8d40d44c43effadb77a351',
+            response_type: 'code',
+            scope: 'offline_access marketplace',
```
Comment (geekgonecrazy) [inline src/misc/cloudAuth.ts]: Note to self and everyone.. before merge we need to set a more restrictive marketplace scope here

File: src/misc/cloudAuth.ts [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+                client_id: '5d8e59c5d48080ef5497e522',
+                code,
+                code_verifier: this.codeVerifier,
+            };
+
+            const url = `${ cloudUrl }/api/oauth/token`;
+            const data = stringify(request);
+            const res = await axios.post(url, data);
+
+            this.config.set('our.a.token', res.data);
```
Comment (geekgonecrazy) [inline src/misc/cloudAuth.ts]: Just reminder res.data will contain:

```
{ access_token: '<access>',
  expires_in: 14400,
  scope: 'offline_access marketplace',
  refresh_token: '<refresh>',
  token_type: 'Bearer' }
```

File: src/misc/cloudAuth.ts [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+    public async hasToken(): Promise<boolean> {
+        await this.initialize();
+
+        return this.config.has('our.a.token');
+    }
+
+    public async getToken(): Promise<string> {
+        await this.initialize();
+
+        return this.config.get('our.a.token', '') as string;
```
Comment (geekgonecrazy) [inline src/misc/cloudAuth.ts]: https://github.com/RocketChat/Rocket.Chat/blob/develop/app/cloud/server/functions/getUserCloudAccessToken.js#L39

here is an example of checking if token is still valid and then refreshing it if not.  Temporarily you can use this secret: `5FEKpI-vcjP-sdmhPi2Z` i've still got to make that route handle with out a secret on the pkce clients

File: src/misc/cloudAuth.ts [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+                            reject(err);
+                        } finally {
+                            this.server.stop();
+                        }
+                    },
+                });
+
+                const codeChallenge = createHash('sha256').update(this.codeVerifier).digest('base64');
+                const authorizeUrl = this.buildAuthorizeUrl(codeChallenge);
+                open(authorizeUrl);
```
Comment (geekgonecrazy) [inline src/misc/cloudAuth.ts]: Maybe we should print it out to the command line to?  Something like:

If your browser didn't open please follow the link below:
<link>

File: src/misc/cloudAuth.ts [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+        }
+    }
+
+    private buildAuthorizeUrl(codeChallenge: string) {
+        const data = {
+            client_id: clientId,
+            response_type: 'code',
+            scope,
+            redirect_uri: this.redirectUri,
+            state: uuidv4(),
```
Comment (geekgonecrazy) [inline src/misc/cloudAuth.ts]: I know we have the code challenge.  But maybe we should also verify state match in the callback as per spec?

Comment (graywolf336) [reply src/misc/cloudAuth.ts]: Is the challenge not good enough?

**Comment by graywolf336** [comment]
@geekgonecrazy login command added.

File: src/misc/cloudAuth.ts (hunk 5 above)
Comment (geekgonecrazy) [reply src/misc/cloudAuth.ts]: Probably good enough.. but we could be better then enough if we wanted :grin: 

Its fine with out it

File: src/misc/cloudAuth.ts [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+            const expiresAt = new Date();
+            expiresAt.setSeconds(expiresAt.getSeconds() + tokenInfo.expires_in);
+
+            this.config.set('rcc.token.access_token', tokenInfo.access_token);
+            this.config.set('rcc.token.expires_in', tokenInfo.expires_in);
+            this.config.set('rcc.token.scope', tokenInfo.scope);
+            this.config.set('rcc.token.token_type', tokenInfo.token_type);
+            this.config.set('rcc.expiresAt', expiresAt);
+        } catch (err) {
+            const d = err.data;
```
Comment (geekgonecrazy) [inline src/misc/cloudAuth.ts]:
```suggestion
            const d = err.response.data;
```

File: src/misc/cloudAuth.ts [hunk 7]
```diff
@@ -0,0 +1,199 @@
...
+            // tslint:disable-next-line:no-console
+            console.log(`[${ err.res.statusCode }] error getting token refreshed: ${ d.error } (${ d.requestId })`);
```
Comment (geekgonecrazy) [inline src/misc/cloudAuth.ts]:
```suggestion
            console.log(`[${ err.response.statusCode }] error getting token refreshed: ${ d.error } (${ d.requestId })`);
```

File: src/misc/cloudAuth.ts [hunk 8]
```diff
@@ -0,0 +1,199 @@
...
+
+            const storageItem: ICloudAuthStorage = {
+                token: tokenInfo,
+                expiresAt,
+            };
+
+            this.config.set('rcc', storageItem);
+
+            return tokenInfo;
+        } catch (err) {
```
Comment (geekgonecrazy) [inline src/misc/cloudAuth.ts]:
```suggestion
        } catch (err) {
            const d = err.response.data;
```

File: src/misc/cloudAuth.ts [hunk 9]
```diff
@@ -0,0 +1,199 @@
...
+            // tslint:disable-next-line:no-console
+            console.log(`[${ err.res.statusCode }] error getting token: ${ err.data.error } (${ err.data.requestId })`);
```
Comment (geekgonecrazy) [inline src/misc/cloudAuth.ts]:
```suggestion
            console.log(`[${ err.response.statusCode }] error getting token: ${ d.error } (${ d.requestId })`);
```

Comment (graywolf336) [reply src/misc/cloudAuth.ts]: You 100% sure about these?

Comment (geekgonecrazy) [reply src/misc/cloudAuth.ts]: Try it.   Alter the refresh token so it’s invalid :)

File: src/misc/cloudAuth.ts [hunk 10]
```diff
@@ -0,0 +1,200 @@
...
+import Conf = require('conf');
+import { createHash } from 'crypto';
+import open = require('open');
+import { stringify } from 'querystring';
+import { cpu, mem, osInfo, system } from 'systeminformation';
+import { v4 as uuidv4 } from 'uuid';
+
+const cloudUrl = 'https://cloud-beta.rocket.chat';
+const clientId = '5d8e59c5d48080ef5497e522';
+const scope = 'offline_access marketplace';
```
Comment (geekgonecrazy) [inline src/misc/cloudAuth.ts]:
```suggestion
const scope = 'offline_access marketplace:submit';
```

**Pull request merged**